File: backend/atlas/liveness/handshake_bind.py
"""Enrolment handshake bind — the "handshake" that welds identity + liveness + co-location
into ONE gesture on ONE timestamp. Stateless, one-shot, nothing stored.

At enrolment the phone shows a RANDOM number N. Holding the phone in the same hand that
wears the ring, the user taps the ring on the phone N times while Face ID authenticates.
Each tap is a sharp mechanical impulse that BOTH IMUs — and the phone microphone — register
at the same instant. That binds three facts to one moment:

  * Face ID    -> WHO   (the identity on the phone)
  * live pulse -> ALIVE (from the ring; checked separately)
  * N co-occurring tap impulses on phone-IMU + ring-IMU (+ mic), at the Face-ID instant
    -> SAME BODY, LIVE, RIGHT NOW  (co-location + a fresh, unreplayable challenge)

`requested_n` is the system's random challenge: the taps must MATCH it, be mutually
time-aligned across channels (one physical impact seen by every sensor), and fall in the
window around the Face-ID success instant. Discrete taps beat a continuous shake — the
phone is still between impacts so Face ID can lock, and counting matched spikes is more
robust than a correlation floor. MEASURES to gate — never a key/value.

HONEST BOUNDARY: needs the ring to stream its IMU. The Colmi R10's accel stream is
intermittent, so on the R10 this degrades to pulse-only and the tap bind can't run — it is
the real thing for an IMU-streaming / secure-element ring. Detection thresholds are tunable
and calibrated per device.
"""
from __future__ import annotations

from typing import List, Optional, Sequence


def detect_taps(signal: Sequence[float], *, fs: float, threshold: float,
                refractory_s: float = 0.12) -> List[float]:
    """Impulse onset times (seconds) from a 1-D motion-magnitude (or audio-energy) signal:
    upward threshold crossings, with a refractory gap so one tap isn't double-counted."""
    times: List[float] = []
    last = -1e9
    dt = 1.0 / fs
    for i in range(1, len(signal)):
        t = i * dt
        if signal[i] >= threshold and signal[i - 1] < threshold and (t - last) >= refractory_s:
            times.append(t)
            last = t
    return times


def _aligned(a: Sequence[float], b: Sequence[float], tol_s: float) -> bool:
    """Bijection: every impulse in `a` has a distinct partner in `b` within `tol_s`."""
    if len(a) != len(b):
        return False
    used = [False] * len(b)
    for ta in a:
        hit = -1
        for j, tb in enumerate(b):
            if not used[j] and abs(ta - tb) <= tol_s:
                hit = j
                break
        if hit < 0:
            return False
        used[hit] = True
    return True


def verify_handshake(*, phone_taps: Sequence[float], ring_taps: Sequence[float],
                     requested_n: int, faceid_at_s: float,
                     mic_taps: Optional[Sequence[float]] = None,
                     window_s: float = 6.0, align_tol_s: float = 0.08) -> bool:
    """True iff the random-N tap challenge was met live and co-located:
      * exactly `requested_n` taps on the phone AND the ring (and mic, if provided),
      * every phone tap co-occurs with a ring tap (and mic tap) within `align_tol_s`
        (one physical impact seen by every channel — same hand / same contact),
      * all taps fall within +/- `window_s` of the Face-ID success instant.
    Fail-closed: any mismatch -> False."""
    if requested_n <= 0:
        return False
    if len(phone_taps) != requested_n or len(ring_taps) != requested_n:
        return False
    lo, hi = faceid_at_s - window_s, faceid_at_s + window_s
    if any(not (lo <= t <= hi) for t in list(phone_taps) + list(ring_taps)):
        return False
    if not _aligned(phone_taps, ring_taps, align_tol_s):
        return False
    if mic_taps is not None:
        if len(mic_taps) != requested_n or not _aligned(phone_taps, mic_taps, align_tol_s):
            return False
    return True

File: ios/AtlasCore/Sources/AtlasCore/Liveness/HandshakeBind.swift
import Foundation

/// Enrolment handshake-bind verifier. Mirrors `backend/atlas/liveness/handshake_bind.py`.
///
/// At enrolment the phone shows a RANDOM number N. Holding the phone in the same hand that
/// wears the ring, the user taps the ring on the phone N times while Face ID authenticates.
/// Each tap is a sharp impulse both IMUs (and the mic) register at the same instant, binding
/// WHO (Face ID) + ALIVE (pulse) + SAME-BODY (co-occurring taps) to one moment. The random N
/// is an unreplayable challenge. Measures to gate — never a key/value.

/// Impulse onset times (seconds) from a 1-D motion-magnitude (or audio-energy) signal:
/// upward threshold crossings, with a refractory gap so one tap isn't double-counted.
public func detectTaps(_ signal: [Double], fs: Double, threshold: Double,
                       refractoryS: Double = 0.12) -> [Double] {
    var times: [Double] = []
    var last = -1e9
    let dt = 1.0 / fs
    guard signal.count > 1 else { return times }
    for i in 1..<signal.count {
        let t = Double(i) * dt
        if signal[i] >= threshold && signal[i - 1] < threshold && (t - last) >= refractoryS {
            times.append(t); last = t
        }
    }
    return times
}

/// Bijection: every impulse in `a` has a distinct partner in `b` within `tolS`.
func handshakeAligned(_ a: [Double], _ b: [Double], tolS: Double) -> Bool {
    guard a.count == b.count else { return false }
    var used = [Bool](repeating: false, count: b.count)
    for ta in a {
        var hit = -1
        for j in 0..<b.count where !used[j] && abs(ta - b[j]) <= tolS { hit = j; break }
        if hit < 0 { return false }
        used[hit] = true
    }
    return true
}

/// True iff the random-N tap challenge was met live and co-located: exactly `requestedN`
/// taps on the phone AND the ring (and mic, if given), each phone tap co-occurring with a
/// ring/mic tap within `alignTolS`, all within ±`windowS` of the Face-ID instant. Fail-closed.
public func verifyHandshake(phoneTaps: [Double], ringTaps: [Double], requestedN: Int,
                            faceIDAtS: Double, micTaps: [Double]? = nil,
                            windowS: Double = 6.0, alignTolS: Double = 0.08) -> Bool {
    guard requestedN > 0 else { return false }
    guard phoneTaps.count == requestedN, ringTaps.count == requestedN else { return false }
    let lo = faceIDAtS - windowS, hi = faceIDAtS + windowS
    if (phoneTaps + ringTaps).contains(where: { !($0 >= lo && $0 <= hi) }) { return false }
    guard handshakeAligned(phoneTaps, ringTaps, tolS: alignTolS) else { return false }
    if let mic = micTaps {
        guard mic.count == requestedN, handshakeAligned(phoneTaps, mic, tolS: alignTolS) else { return false }
    }
    return true
}
